java/fury-core/src/main/java/org/apache/fury/reflect/TypeToken.java

```diff
@@ -0,0 +1,330 @@
+ * the License.
+ */
+
+package org.apache.fury.reflect;
+
+import com.google.common.reflect.TypeParameter;
+import java.lang.reflect.Array;
+import java.lang.reflect.GenericArrayType;
+import java.lang.reflect.ParameterizedType;
+import java.lang.reflect.Type;
+import java.lang.reflect.TypeVariable;
+import java.lang.reflect.WildcardType;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.Set;
+
+public class TypeToken<T> {
+
+  private final Type type;
+
+  /**
+   * Constructs a new type token of {@code T}.
+   *
+   * <p>Clients create an empty anonymous subclass. This embeds the type parameter in the anonymous
+   * class's type hierarchy, so we can reconstitute it at runtime despite erasure.
+   *
+   * <p>For example:
+   *
+   * <pre>{@code
+   * TypeToken<List<String>> t = new TypeToken<List<String>>() {};
+   * }</pre>
+   */
+  protected TypeToken() {
+    this.type = capture();
+  }
+
+  private TypeToken(Class<T> declaringClass) {
+    this.type = declaringClass;
+  }
+
+  private TypeToken(Type type) {
+    this.type = type;
+  }
+
+  /** Returns an instance of type token that wraps {@code type}. */
+  public static <T> TypeToken<T> of(Class<T> clazz) {
+    return new TypeToken<T>(clazz);
+  }
+
+  /** Returns an instance of type token that wraps {@code type}. */
+  public static <T> TypeToken<T> of(Type type) {
+    return new TypeToken<T>(type);
+  }
+
+  /** Returns the captured type. */
+  private Type capture() {
+    final Type superclass = getClass().getGenericSuperclass();
+    if (!(superclass instanceof ParameterizedType)) {
+      throw new IllegalArgumentException(superclass + " isn't parameterized");
+    }
+    return ((ParameterizedType) superclass).getActualTypeArguments()[0];
+  }
+
+  /** Returns the represented type. */
+  public Type getType() {
+    return type;
+  }
+
+  /**
+   * Returns the raw type of {@code T}. Formally speaking, if {@code T} is returned by {@link
+   * java.lang.reflect.Method#getGenericReturnType}, the raw type is what's returned by {@link
+   * java.lang.reflect.Method#getReturnType} of the same method object. Specifically:
+   *
+   * <ul>
+   *   <li>If {@code T} is a {@code Class} itself, {@code T} itself is returned.
+   *   <li>If {@code T} is a {@link ParameterizedType}, the raw type of the parameterized type is
+   *       returned.
+   *   <li>If {@code T} is a {@link GenericArrayType}, the returned type is the corresponding array
+   *       class. For example: {@code List<Integer>[] => List[]}.
+   *   <li>If {@code T} is a type variable or a wildcard type, the raw type of the first upper bound
+   *       is returned. For example: {@code <X extends Foo> => Foo}.
+   * </ul>
+   */
+  public Class<? super T> getRawType() {
+    // For wildcard or type variable, the first bound determines the runtime type.
+    Class<?> rawType = getRawTypes(type).iterator().next();
```

Comment:
This may be expensive. Fury init will invoke this API many times.

Reply:
I will rewrite and update the patch.